moma_mission/src/moma_mission/states/observation.py

```diff
@@ -66,3 +67,138 @@ def run(self):
+        self.controlled_frame = self.get_scoped_param("controlled_frame", "controlled_frame")
+
+        # the approximate lumped size of the object bounding box
+        self.object_size = self.get_scoped_param("object_size", 1.0)
+
+        # how big in the image the object should look like. 1.0 means getting the closest to the object
+        self.image_fill_ratio = self.get_scoped_param("image_fill_ratio")
+
+        # needed for FOV computation
+        self.camera_info_topic = self.get_scoped_param("camera_info_topic", "/camera_info")
+        self.heading = self.get_scoped_param("heading", [1, 0, 0])
+        
+        # how much to look lateraly across all candiate observation poses (it is the phi angle in sperical coords)
+        self.lateral_view_angle = np.deg2rad(self.get_scoped_param("lateral_view_angle", 20))
+        
+        self.sample_poses = []
+        self.num_samples = 1 + 8  # central  + intermediate angles
+        
+        self.attempts = 0
+        self.first_run = True
+        
+        self.camera_info = None
+        self.camera_info_sub = rospy.Subscriber(self.camera_info_topic, CameraInfo, self._camera_info_cb)
+        self.pose_array_pub = rospy.Publisher("/sampled_viewpoints", PoseArray, queue_size=1)
+        self.pose_broadcaster = tf2_ros.StaticTransformBroadcaster()
+
+    def _spherical_to_cartesian(self, r, theta, phi):
+        r_sin_phi = r * np.sin(phi)
+        return np.array([r_sin_phi * np.cos(theta),
+                         r_sin_phi * np.sin(theta),
+                         r * np.cos(phi)])
+
+    def _camera_info_cb(self, msg):
+        self.camera_info = msg
+        
+    def run(self):
+        if not self.camera_info:
+            rospy.logerr("No camera_info message received on {}".format(self.camera_info_topic))
+            return 'Failure'
+
+        if self.first_run:
+            w = np.min([self.camera_info.height, self.camera_info.width])
+            f = np.max([self.camera_info.K[0], self.camera_info.K[4]])
+            d = self.object_size * f / ( w * self.image_fill_ratio)
+
+            T_map_obj = self.get_transform(self.map_frame, self.object_frame)
+            T_cam_ctrl = self.get_transform(self.camera_frame, self.controlled_frame)
+        
+            thetas = [0.0] + [i * np.pi/4 for i  in range(8)]
+            phis = [0.0] + [self.lateral_view_angle] * 8
+
+            sample_pose_array = PoseArray()
+            for theta, phi in zip(thetas, phis):
+                coords = self._spherical_to_cartesian(d, theta, phi)
+
+                z_axis_dir = -tf.transformations.unit_vector(coords)
+                x_axis_dir_desired = self.heading
+                y_axis_dir = tf.transformations.unit_vector(np.cross(z_axis_dir, x_axis_dir_desired))
+                x_axis_dir = np.cross(y_axis_dir, z_axis_dir)
+                assert np.allclose(x_axis_dir, tf.transformations.unit_vector(x_axis_dir))
```

Comment:
Do you think it is feasible to inherit from `SphericalSamplerState` (maybe outsource some functionality to separate functions there that can be reused) and avoid some duplication?